test/distributed/_tools/test_mem_tracker.py

```diff
@@ -0,0 +1,234 @@
+            output = model(input_batch)
+            output.sum().backward()
+            output = None
+            optim.step()
+            optim.zero_grad()
+
+        # Check for accuracy of peak memory
+
+        tracker_max = mt.get_tracker_snapshot("peak")[dev]["Total"]
+        cuda_max = torch.cuda.max_memory_allocated(dev)
+        accuracy = tracker_max / cuda_max
+
+        self.assertAlmostEqual(accuracy, 1.0, delta=0.1)
+
+    @skipIfTorchDynamo("https://github.com/pytorch/pytorch/issues/115653")
+    @onlyCUDA
+    @parametrize(
+        "bsz, n_layers, dim, dtype",
+        [(128, 4, 1024, torch.bfloat16)],
+    )
+    def test_tracker_with_activation_checkpointing(
+        self,
+        device: str,
+        bsz: int,
+        n_layers: int,
+        dim: int,
+        dtype: torch.dtype,
+    ):
+        """
+        Tests that the tracker correctly computes the peak memory during activation checkpointing.
+        """
+        dev = torch.device(device)
+        torch.cuda.reset_accumulated_memory_stats()
+        torch.cuda.reset_peak_memory_stats()
+
+        class MLPBlock(nn.Module):
+            def __init__(self, dim: int):
+                super().__init__()
+                self.mlp_block = nn.Sequential(
+                    nn.Linear(dim, 2 * dim), nn.ReLU(), nn.Linear(2 * dim, dim)
+                )
+
+            def forward(self, x):
+                return self.mlp_block(x)
+
+        class MyModule(nn.Module):
+            def __init__(self, n_layers: int, dim: int, use_ac: bool = False):
+                super().__init__()
+                self.mlp_blocks = nn.ModuleList()
+                self.use_ac = use_ac
+                for _ in range(n_layers):
+                    self.mlp_blocks.append(MLPBlock(dim))
+
+            def forward(self, x):
+                for i, block in enumerate(self.mlp_blocks):
+                    if i >= 1 and self.use_ac:
+                        x = checkpoint(
+                            block, x, preserve_rng_state=True, use_reentrant=False
+                        )
+                    else:
+                        x = block(x)
+                return x
+
+        with torch.device(dev):
+            model = MyModule(n_layers, dim, True).to(dtype=dtype)
+        optim = torch.optim.Adam(model.parameters(), foreach=True)
+        mem_tracker = MemTracker()
+        mem_tracker.track_external(model, optim)
+        with mem_tracker as mt:
+            input_batch = torch.randn(bsz, dim, dim, device=dev, dtype=dtype)
+            output = model(input_batch)
+            output.sum().backward()
+            optim.step()
+            optim.zero_grad()
+
+        # Check for accuracy of peak memory
+        tracker_max = mt.get_tracker_snapshot("peak")[dev]["Total"]
+        cuda_max = torch.cuda.max_memory_allocated(dev)
+        accuracy = tracker_max / cuda_max
+        self.assertAlmostEqual(accuracy, 1.0, delta=0.1)
```

Comment:
Do we care if we are over-estimating or under-estimating?

Reply:
The delta will always be under-estimation. Essentially all the tensors, that go through the dispatch mode will be caught. But I have seen a few places in PyTorch that call some operations under no-dispatch, so that memory won't be captured.